### DISCUSSION TITLE
Unmounting mounted app

### DISCUSSION THREAD

**Discussion by matan1008** [encode/starlette#2028] [Ideas]
Hey!
I thought it might be cool (and useful in my case) to unmount mounted apps.
Would be greate if you could check https://github.com/encode/starlette/pull/2027

Thanks a lot!

**Comment by Kludex** [comment]
What's the use-case?

**Reply by matan1008** [reply]
In my case, I want to allow admin to load and unload users sub application (dynamically).
Loading applications is easy with the mount, unloading applications is not yet implemented

**Reply by alex-oleshkevich** [reply]
@matan1008, you can go with `Mount` subclass and see how it works.

**Reply by matan1008** [reply]
That's what I did to load sub application, how do you unload it?

**Reply by alex-oleshkevich** [reply]
can you share some code snippet? so we can better understand your case.

**Reply by matan1008** [reply]
Sure!

```python
app = FastAPI()
subapps_manager = SubappsManager()


# Adding a subapp works fine
@app.post('/subapps/{name}')
async def add_subapp(request: Request, name: str):
    subapp = subapps_manager.get_subapp(name)
    app.mount(f'/{name}', subapp)
    return JSONResponse({})


# Deleting subapp missing API in my opinion
@app.delete('/subapps/{name}')
async def delete_subapp(request: Request, name: str):
    ...

```

My suggestion (and the PR that follows):

```python
app = FastAPI()
subapps_manager = SubappsManager()
mounted_apps = {}


@app.post('/subapps/{name}')
async def add_subapp(request: Request, name: str):
    subapp = subapps_manager.get_subapp(name)
    mounted_apps[name] = app.mount(f'/{name}', subapp)
    return JSONResponse({})


@app.delete('/subapps/{name}')
async def delete_subapp(request: Request, name: str):
    app.unmount(mounted_apps[name])
    del mounted_apps[name]
    return JSONResponse({})

```

**Reply by alex-oleshkevich** [reply]
Thanks! This is a very untypical use case. Not sure it this should be a part of Starlette. 

Alternatively you can:
1. subclass `Mount` and modify `app.routes.routes` 
2. or, have a middleware that properly routes users to a specific route branches

I would personally go with 2 as the most flexible and error-prone solution.

**Reply by matan1008** [reply]
Thanks a lot!
I am facing a similar problem - how to remove an added middleware?

**Reply by alex-oleshkevich** [reply]
I mean, you don't need to remove it. In the middleware, check who is the user and allow/deny access to the sub route.

It may look like this:

```python
def can_access(user, subapp) -> bool:
    """some logic to test if user can access subapp"""

def GuardMiddleware:
    def __init__(self, app):
        self.app = app

    async def __call__(self, scope, receive, send):
        request = Request(scope, receive, send)

        if 'subapps/admin' in request.url.path and can_access(request.user, 'admin'):
            return await self.app(scope, receive, send) # grant

        if 'subapps/anotherapp' in request.url.path and can_access(request.user, 'anotherapp'):
            return await self.app(scope, receive, send) # grant

		response = PlainTextResponse('Forbidden', 403)
        await response(scope, receive, send) # forbid
```

**Reply by matan1008** [reply]
Sorry, I didn't explain the use case well enough. First, I want load the sub application dynamically (literally, using importlib) - in my example the `get_subapp` would return different app each time. Second, after loading the app all users should be able to access it and after removing it no one should be able to access it
